shared/src/main/java/com/faforever/neroxis/util/ImageUtil.java

```diff
@@ -135,14 +135,121 @@ public static void writeNormalDDS(BufferedImage image, Path path) throws IOExcep
+            channel = 2;
+        }
+        return writeTextureIntoImage(baseImage, heightMap, layer, channel);
+    }
 
-    public static byte[] getMapwideTextureBytes(NormalMask normalMask, FloatMask shadowMask) {
+    public static BufferedImage setRoughnessMap(BufferedImage baseImage, BufferedImage roughnessMap, int layer) {
+        if (roughnessMap.getColorModel().getNumComponents() != 1) {
+            throw new IllegalArgumentException("Roughness map needs to be greyscale.");
+        }
+        if (baseImage.getColorModel().getNumComponents() != 4) {
+            throw new IllegalArgumentException("base image needs to be RGBA.");
+        }
+        if (baseImage.getWidth() != roughnessMap.getWidth() * 4) {
+            throw new IllegalArgumentException((
+                    "Texture sizes are not compatible: base image width %d, roughness map width %d. " +
+                            "The roughness map needs to be quarter the resolution of the base image."
+            ).formatted(baseImage.getWidth(), roughnessMap.getWidth()));
+        }
+        int channel = 1;
+        if (layer >= 4) {
+            channel = 3;
+        }
+        return writeTextureIntoImage(baseImage, roughnessMap, layer, channel);
+    }
+
+    private static BufferedImage writeTextureIntoImage(BufferedImage baseImage, BufferedImage texture, int layer, int channel) {
+        int offsetX = baseImage.getWidth() / 2;
+        int offsetY = baseImage.getHeight() / 2;
+        int textureWidth = texture.getWidth();
+        int textureHeight = texture.getHeight();
+        WritableRaster baseImageRaster = baseImage.getRaster();
+        // We need to write the texture with padding. We can achieve that by offsetting it and writing it in a 2x2 grid
+        texture = offsetHalf(texture);
+        Raster imageRaster = texture.getData();
+        for (int a = 0; a <= 1; a++) {
+            for (int b = 0; b <= 1; b++) {
+                for (int x = 0; x < textureWidth; x++) {
+                    for (int y = 0; y < textureHeight; y++) {
+                        int newX = x + a * textureWidth;
+                        int newY = y + b * textureHeight;
+                        if (layer % 2 == 1) {
+                            newX += offsetX;
+                        }
+                        if (layer % 4 >= 2) {
+                            newY += offsetY;
+                        }
+                        int value = imageRaster.getPixel(x, y, new int[1])[0];
+                        if (texture.getType() == BufferedImage.TYPE_USHORT_GRAY) {
+                            value /= 256;
+                        }
+                        int[] baseImagePixel = baseImageRaster.getPixel(newX, newY, new int[4]);
+                        baseImagePixel[channel] = value;
+                        baseImageRaster.setPixel(newX, newY, baseImagePixel);
+                    }
+                }
+            }
+        }
+        return baseImage;
```

Comment:
Is this just writing masks into images with some offset? Util methods already exist for this kind of operation. So I guess need a little more insight onto why this needs it's own special method. 

It is also named as though it is general when there is some very special operations going on here.

Reply:
Yes, it is writing into the image with an offset that I calculate based on the layer number. I need to control which channel we are writing into because I need to write into one channel only per mask. The `insertImageIntoNewImageOfSize` method looked like it would write a greyscale image into all the channels of the bigger image.

Good catch that it should not have such a generic name.

Reply:
It isn't only grayscale, or shouldn't be it just depends on the image types themselves. And it can can larger or smaller or offset.

Also you could instead just create a vector 4 mask as well and then save that as an image using normal utilities.

Reply:
Do you think doing these operations on a mask would be simpler?

Reply:
Yes because then you could work easily on individual layers instead of all the conditional checks about which layer are we actually working on.

Reply:
And the offset methods already exist rather than having another version.

Ideally all processing is done on masks and then persistence is done in images. It keeps the logic and abstraction layers simpler.

Reply:
doing it directly with images seemed easy because I already load images.
Which methods would I use to set a single component of the Vector4Mask? And how can I copy a smaller mask into a bigger mask with offset? I couldn't find any suitable methods.

Reply:
There are the withOffset methods and Component methods that can be used.

Reply:
I'm afraid I can't really wrap my head around the stuff that goes on in the masks. I started using them for the processing, but I get index out of bounds errors. 2048 is obviously out of bounds when the array is 2048 elements long, but I can't figure out why the index grows to that, the loop method seems correct. You can have a look by running the pbr texture generator.
I also have to write a function that writes the vector mask into an image, but that is beyond my abilities.

Reply:
I can take a look in maybe a few days. Also there is a method that writes a vector mask to an image already. Does that not work?

Reply:
No rush, it's not urgent.
That method does some processing on the values, which are probably couterproductive for what I want to do, but I would have to look closer.
I have been thinking again about the approach to do the processing in masks. I can see the use when you start out with masks, then it makes sense to keep all the processing in the masks and only to convert on the last step.
But in this case we start out with images, and we don't do any processing on the values, just writing the values into a different area. When looking into the masks I noticed that I would have to write new methods for some of the steps I need to do anyway and it's 10 times harder for me because the masks are doing a lot of pretty abstract stuff that I need to understand first before I can make a call what methods to use or how to write my new methods.
On the contrary the approach with the image manipulation works already. I wonder if it's really worth it to basically reimplement the whole thing to satisfy the abstract notion of "processing should be done in masks". In my personal opinion it isn't, partly because that rewrite would take significantly longer than fixing my first implementation would take.

I concede that my first implementation is not the most readable, but I think this can be refactored relatively easily. Maybe it is also a good idea to not store the methods in imageUtil, but in the tool directly, as they are pretty special in what they do and not really suitable for other use cases.

Reply:
My main thinking is that I don't want the processing aspects to be split into multiple places. ImageUtil is really for basic operations centered around converting masks to images.

I more so want to stay away from having duplicate operations one for masks and one for images as that can be a maintenance pain and also with images there are a lot of caveats that one would have to make assumptions about which makes it less applicable and easier to fall into traps.

So that is why I prefer processing is done in masks because there are less assumptions about what the is required and the then can be easily translated.

And what steps would you need new methods for?

Reply:
I thought I needed to write a new method for writing the vector4Mask to an image without processing, because toImage() does a lot of processing, but I now found writeToImage() from the vectorMask class.
I guess it would also be nicer to use setWithOffset instead of addWithOffset, because that is closer to what is supposed to happen, but it doesn't matter when performing this operation on a floatMask that is 0 everywhere.

Reply:
Can you try to run the PbrTextureGenerator and help me with that out of bounds error? I don't know why this happens. If that error is overcome then I think I can do the rest.

Reply:
We can add setWithOffset it should be there. I probably just got lazy

Reply:
And yeah sorry guess I should have been more specific that method already existed